### PR TITLE
patmat: Break up enumerateSubtypes

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by SethTisue** [comment]
review by @martijnhoekstra, perhaps?

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 1]
```diff
@@ -81,94 +81,71 @@ trait TreeAndTypeAnalysis extends Debugging {
...
-              val memberType = nestedMemberType(sym, pre, tpApprox.typeSymbol.owner)
-              val subTp = appliedType(memberType, WildcardType.fillList(sym.typeParams.length))
-              val subTpApprox = analyzer.approximateAbstracts(subTp) // TODO: needed?
-              // debug.patmat("subtp"+(subTpApprox <:< tpApprox, subTpApprox, tpApprox))
-              if (subTpApprox <:< tpApprox) Some(checkableType(subTp))
-              else None
-            }
-          }
+    def enumerateSubtypes(tp: Type, grouped: Boolean): List[List[Type]] = tp.typeSymbol match {
+      case    UnitClass                 => List(List(UnitTpe))
```
Comment (NthPortal) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: kind of weird alignment

```suggestion
      case UnitClass                    => List(List(UnitTpe))
```

**Comment by NthPortal** [comment]
Dale, your commit message 😂🤣

**Review by lrytz — APPROVED** [review]
https://github.com/scala/scala/commit/d44a86f432 for reference where the `!grouped` guards originate

**Comment by lrytz** [comment]
squash the last commit?

**Comment by martijnhoekstra** [comment]
Where the `!grouped` came from is easier determined than what its intent was, and whether we now undermine that intent. There are two places where the method is called with grouped = false and one where grouped = true, but I haven't gotten to the bottom of when and why one is done over the other (and why the `!grouped` guards would be valid or not).

At the very least, this PR makes figuring that out *much* easier than it was before.

**Comment by dwijnand** [comment]
> grouped = true, but I haven't gotten to the bottom of when and why one is done over the other (and why the `!grouped` guards would be valid or not)

My changes makes more unit, boolean and modules return their subtypes (in 1 group) even when looking for grouped subtypes.  Grouped subtypes, as you say, is used once, to define groupedDomains which in turn is used twice:

The second usage of groupedDomain is guarded behind `syms.size > 1` which means at most (lol) it will add an axiom of `AtMostOne(ConstantTrue :: ConstantFalse :: Nil)`, one per variable in the match expression that is of type boolean, which sounds fine.

The first usage of groupedDomain is preserving the semantics of the cases that that optimisation doesn't cover: for non-exclusive (aka AtMostOne) domains, use the old strategy of adding the "sym implies !excludedSym" axiom (as `Or(Not(sym), Not(excludedSym))`, aka `!sym || !excludedSym` which is the normalised form of the, IMHO, easier to grok `!(sym && excludedSym)`).  You need more than one subtype for any one symbol to have an exclude implication on the other, so again we're only talking about boolean's false + true subtypes.  So for booleans we now no longer add the `!(true && false)` and `!(false && true)` axioms, which works out fine because we've replaced that with the adding of `AtMostOne(true, false)` above.  So I guess what I've done is extended that optimisation to booleans too.

Perhaps I'm mistaken somewhere above, but I kind of assume a test or bootstrapping would break if I'm breaking something here.  And, either way, even if something is breaking and not showing I don't think the solution is "dear God, if you're asking for grouped subtypes, well then Boolean is uninhabited!" and would rather fix the problem properly.

My guess is the `!grouped` guards were added in a kind of "I'm here to help optimise the explosion of sealed hierarchy variables, so not any of the other branches here".

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 2]
```diff
@@ -81,94 +81,71 @@ trait TreeAndTypeAnalysis extends Debugging {
...
+        case Nil      => acc
+        case hd :: tl =>
+          val children = filterAndSortChildren(hd.sealedChildren)
+          // put each trait in a new group since traits could belong to the same group as a derived class
+          val (traits, nonTraits) = children.partition(_.isTrait)
+          groupChildren(tl ::: children, acc ::: traits.map(List(_)).appended(nonTraits))
+      }
+
+      val subclasses = debug.patmatResult(s"enum $sym sealed, subclasses") {
+        if (grouped) groupChildren(List(sym), Nil)
```
Comment (martijnhoekstra) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:139]: This is the only remaining use of grouped. Assuming both branches return the same elements but grouped differently, is it worth it to eliminate grouped altogether and always make it true? If so, there are two call sites that call it with false, of which one would have to be adapted not to use headOption

Comment (dwijnand) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:139]: Yeah, I wanted to too.  But I didn't want to get into trouble with the allocation police.

Comment (martijnhoekstra) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:139]: The allocation police will eventually come by and change those `:::`'s to mutable builders anyway

**Review by martijnhoekstra — APPROVED** [review]
This is so much cleaner and understandable than it was, thanks!

**Review by martijnhoekstra — APPROVED** [review]

**Pull request merged**